### DISCUSSION TITLE
Creating a generic for an unknown number of argument of a Callable in python

### DISCUSSION THREAD

**Discussion by rayannott** [community/community#77313] [Programming Help] [answered]
### Body

In this programming exercise, I wrote a starmap generator as an equivalent to itertools.starmap and aimed to provide accurate type hints. The code snippet below showcases the implementation:

```python
from typing import Callable, Generator, Iterable, Unpack, TypedDict


def starmap[T](func: Callable[..., T], iterable: Iterable[tuple]) -> Generator[T, None, None]:
    yield from map(func, *zip(*iterable))

# test:
import itertools
args = lambda x, y: x + y, [(1, 2), (3, 4)]
assert list(itertools.starmap(*args)) == list(starmap(*args))
```

The challenge now lies in determining how to generically represent the arguments (...) of the func parameter so that their types align with those found in the tuples within the provided iterable. How does one write the generic for the arguments?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by CheetahCodes21** [comment]
The generic for the arguments should be represented using a `Tuple[Any, ...]`. This indicates that the `func` parameter accepts any number of arguments, each of which can be of any type. The `Any` type is a placeholder for unknown or generic types.

In the provided code snippet, the `func` parameter is defined as `Callable[..., T]`, where `T` is the return type of the function. This means that `func` can take any number of arguments and return a value of type `T`. However, the specific types of the arguments are not specified.

To specify the types of the arguments, you can use the `Tuple[Any, ...]` type. This type indicates that the `func` parameter takes a tuple of any number of arguments, each of which can be of any type. For example, if you want the `func` parameter to take a tuple of two arguments, you can use the following type:

```python
func: Callable[[Tuple[Any, Any]], T]
```

This type indicates that `func` takes a tuple of two arguments, where the first argument can be of any type and the second argument can be of any type. The return type of `func` is still `T`.

By using the `Tuple[Any, ...]` type, you can ensure that the types of the arguments in the `iterable` parameter align with the types of the arguments that the `func` parameter expects.

Here is an example of how to use the `starmap` function with a generic type for the arguments:

```python
def starmap_typed[T](func: Callable[..., T], iterable: Iterable[tuple]) -> Generator[T, None, None]:
    # Check that the number of arguments in each tuple matches the number of arguments expected by the function
    for args in iterable:
        if len(args) != len(func.__annotations__.keys()):
            raise ValueError(f"Function {func.__name__} expects {len(func.__annotations__.keys())} arguments, but got {len(args)}")

    # Check that the types of the arguments in each tuple match the expected types
    for args in iterable:
        for i, arg in enumerate(args):
            if type(arg) != func.__annotations__[f"arg{i}"]:
                raise ValueError(f"Expected argument {i} of type {func.__annotations__[f'arg{i}']}, but got {type(arg)}")

    # Call the function with the arguments
    yield from map(func, *zip(*iterable))
```

This version of the `starmap` function checks that the number of arguments in each tuple matches the number of arguments that the `func` parameter expects. It also checks that the types of the arguments in each tuple match the expected types. If there is a mismatch, an error is raised.

I hope this helps!

**Reply by rayannott** [reply]
@CheetahCodes21, thanks for your answer. However, I mostly asked about how to type it so that the tuple of types in the `Callable[here, T]` matches the tuple in the interable `Iterable[here]`.
The result of such type hinting would be the type checking error returned by the following code:
```python
def func(a: int, b: list[int]) -> list[int]:
    return a * b

it = starmap(func, [(3, ['1', '2']), (2, [4, 5, 6])])
``` 
It should've checked that `['1', '2']` is not `list[int]`, but now it doesn't care.

**Reply by CheetahCodes21** [reply]
Hey @rayannott ,
You're correct that the current type hinting for `starmap` doesn't enforce that the elements of the inner iterables match the expected types of the function's arguments. This is because the `Callable` type parameter in the signature of `starmap` only specifies the overall return type of the function, not the types of its arguments.

To achieve the desired type checking behavior, you'll need to use a custom type annotation for the iterable argument of `starmap`. One approach is to define a custom generic type that encapsulates the expected types of the function's arguments and the return type. Here's an example of how to do this:

```python
from typing import Iterable, Callable, TypeVar

T = TypeVar('T')

def starmap_with_type_checking(func: Callable[..., T], iterable: Iterable[tuple[Any, ...]], return_type: T) -> Iterable[T]:
    for args in iterable:
        yield func(*args)
```

This custom `starmap_with_type_checking` function takes three type parameters:

1. `func`: The callable to be applied to each element of the iterable.

2. `iterable`: The iterable of tuples, where each tuple represents the arguments for a single call to the function.

3. `return_type`: The expected return type of the function.

By providing the `return_type` parameter, you can enforce type checking for the result values of the function calls.

To use this custom function, you would annotate the iterable argument with the custom type:

```python
it = starmap_with_type_checking(func, [(3, ['1', '2']), (2, [4, 5, 6])], list[int])
```

This type annotation ensures that the type checker will raise an error if any of the inner iterables contain elements that don't match the expected types of the function's arguments. For example, if you change the second element of the first tuple to a non-list value like `"string"`, the type checker will report an error.

**Reply by rayannott** [reply]
Hmm, yes, this looks like a valid solution, thank you.
Although, I would probably not use it as it requires passing another argument. I much rather do explicit runtime checks like the ones that you suggested [here](https://github.com/orgs/community/discussions/77313#discussioncomment-7705999).

Upd.: I just tried running the code as inteded, and pylance still shows an error.

Code:
```python
from typing import Iterable, Callable, TypeVar

T = TypeVar('T')

def starmap_with_type_checking(func: Callable[..., T], iterable: Iterable[tuple[Any, ...]], return_type: T) -> Iterable[T]:
    for args in iterable:
        yield func(*args)

it = starmap_with_type_checking(func, [(3, [1, 2]), (2, [4, 5, 6])], list[int])
```

The error:
```
Argument of type "list[tuple[Literal[3], list[int]] | tuple[Literal[2], list[int]]]" cannot be assigned to parameter "iterable" of type "Iterable[tuple[Any, ...]]" in function "starmap_with_type_checking"
  "Literal[3]" is incompatible with "Any"
  "list[int]" is incompatible with "Any"
  "Literal[2]" is incompatible with "Any"
  "list[int]" is incompatible with "Any"Pylance
```

**Reply by CheetahCodes21** [reply]
The error message indicates that the type of the `iterable` argument in the call to `starmap_with_type_checking` is incompatible with the expected type. The expected type is `Iterable[tuple[Any, ...]]`, which means an iterable of tuples where each tuple can have any type of element. However, the provided iterable is `[(3, [1, 2]), (2, [4, 5, 6])]`, which is a list of tuples where each tuple has two literal integer elements followed by a list of integers.

To fix the error, you need to modify the type annotation for the `iterable` parameter in the `starmap_with_type_checking` function to reflect the actual type of the iterable being passed. In this case, the type annotation should be `Iterable[tuple[Literal[3, 2], List[int]]]`. This will inform Pylance that the iterable contains tuples where the first element is either a literal 3 or a literal 2, and the second element is a list of integers.

Here's the updated code with the corrected type annotation:

```python
from typing import Iterable, Callable, TypeVar

T = TypeVar('T')

def starmap_with_type_checking(func: Callable[..., T], iterable: Iterable[tuple[Literal[3, 2], List[int]]], return_type: T) -> Iterable[T]:
    for args in iterable:
        yield func(*args)

it = starmap_with_type_checking(func, [(3, [1, 2]), (2, [4, 5, 6])], list[int])
```

With this change, Pylance should no longer show an error.

**Reply by rayannott** [reply]
Well, of course, one can add such type annotation, but then the whole point of writing them vanishes. If one changes the `func` or the structure of the input arguments, the problem will come back.

My goal is to keep the definition of `starmap` as general as possible and still get warnings from the type checker.

**Reply by CheetahCodes21** [reply]
I understand your concern about maintaining the generality of the `starmap` function while still receiving warnings from the type checker. In this case, the issue arises from the specific structure of the input arguments being passed to `starmap`. The type checker is unable to infer the exact types of the elements in the tuples because they are dynamically constructed within the call to `starmap`.

To address this issue, you could consider utilizing a more specialized function for handling tuples of literal integers followed by lists of integers. This specialized function could have its own type annotations that explicitly specify the expected types of the input arguments. This approach would allow you to maintain the generality of the `starmap` function while still receiving type checking for the specific use case.

Alternatively, you could consider using a third-party type annotation library, such as `mypy-extensions`, which provides additional type annotations and features that can aid in type checking more complex code structures. These libraries may offer more granular control over type checking and allow you to achieve the desired level of specificity without sacrificing generality.

Ultimately, the choice of approach depends on the specific requirements of your codebase and the desired balance between generality and type checking.

**Reply by rayannott** [reply]
Okay, I see.

So, in the end, there is no simple way to annotate the arguments of the function so that
- they match the types in the tuple from `Iterable` and
- there is an arbitrary number of them with arbitrary types?

**Reply by CheetahCodes21** [reply, marked as answer]
Yes, that's correct. The inherent flexibility of Python's dynamic type system makes it challenging to achieve precise type checking for functions like `starmap` that handle arbitrary numbers of arguments with varying types. While type annotations provide valuable information for type checkers, they cannot fully capture the dynamic nature of such functions.

In situations where precise type checking is crucial for specific use cases, consider utilizing specialized functions with explicit type annotations tailored to those cases. This approach maintains generality for the main function while ensuring type safety for the specific scenarios that require it.

Additionally, explore third-party type annotation libraries like `mypy-extensions` that offer more granular control over type checking and may provide solutions for handling complex type structures.

Ultimately, the choice of approach depends on the specific requirements of your codebase and the desired balance between generality and type checking.

**Comment by CheetahCodes21** [comment]
To generically represent the arguments (...) of the `func` parameter, you can utilize type inference and generic type parameters. Here's the modified code with generics:

```python
from typing import Callable, Generator, Iterable, TypeVar, Unpack, TypedDict


def starmap[T](func: Callable[..., T], iterable: Iterable[tuple]) -> Generator[T, None, None]:
    Element = Unpack[tuple[...]]
    ElementType = TypeVar('ElementType', bound=Element)

    def _starmap(element: ElementType) -> T:
        return func(*element)

    return map(_starmap, iterable)

# test:
import itertools
args = lambda x, y: x + y, [(1, 2), (3, 4)]
assert list(itertools.starmap(*args)) == list(starmap(*args))
```

In this code, `ElementType` is a type variable that represents the element type of the tuples in the `iterable` parameter. The `_starmap` function unpacks the tuple and passes the individual elements to the `func` function. This ensures that the types of the arguments passed to `func` align with the types found in the tuples within the `iterable`.

The `TypeVar` annotation on `ElementType` specifies that this type variable is bound to `Element`, which represents the type of a single element in an unpacked tuple. This ensures that the `_starmap` function can only be called with elements that are compatible with the types found in the `iterable`.

By using type inference and generic type parameters, you can effectively generically represent the arguments of the `func` parameter and ensure type safety when using the `starmap` function.
